### DISCUSSION TITLE
How to use WinRT in Avalonia?

### DISCUSSION THREAD

**Discussion by DmitryBorodiy** [AvaloniaUI/Avalonia#15799] [Q&A] [answered]
Hello, Avalonia community!
I need to use WinRT inside Avalonia app(on Windows only) and at the same time, stay cross-platform. Any suggestions?

Thank you in advance!
If you can please provide code example🙏

**Comment by timunie** [comment]
1. Add an interface into main project
2. Add a Windows project and other platfrom project
3. Implement the Interface of 1) into each platform project with the needed API

**Comment by maxkatz6** [comment]
To add to @timunie comment, to access WinRT APIs you can use [CsWinRT](https://github.com/microsoft/CsWinRT) library.
OR (probably easier) just use specific windows TargetFramework, like `<TargetFramework>net6.0-windows10.0.19041.0</TargetFramework>` and then you should be able to access APIs from this Windows version. See https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/wpf-plus-winappsdk (it's not limited by WPF)

**Reply by DmitryBorodiy** [reply]
What if I want the app to work on Linux and Mac, but only enable WinRT on Windows?

**Reply by thevortexcloud** [reply]
Using dependency injection with multiple projects is a good approach then. Otherwise you can multi target, but it's fairly annoying to manage.

**Reply by DmitryBorodiy** [reply]
I can use something like
`<TargetFramework>net6.0-windows10.0.19041.0, net6.0</TargetFramework>`  to run on Linux also? Can you show me example solution or project source code?

**Reply by DmitryBorodiy** [reply]
I just want to use Windows-specific APIs such as Share Dialog without losing cross-platform ability.

**Reply by J0nathan550** [reply]
> I just want to use Windows-specific APIs such as Share Dialog without losing cross-platform ability.

[FluentAvaloniaTheme](https://amwx.github.io/FluentAvaloniaDocs/pages/FATheme)

**Reply by thevortexcloud** [reply, marked as answer]
You can use Xamarin essentials for that. It already implements the share dialogue for most platforms.

https://learn.microsoft.com/en-us/previous-versions/xamarin/essentials/share?tabs=uwp

With that said, that is not the way you multi target. Instead you do:

```xml
<TargetFrameworks>net6.0-windows10.0.19041.0;net6.0</TargetFrameworks>
```

When you build the project, it will output assemblies for both targets. The problem with multi targeting (at least with platform specific TFMs) is that you can't have any reference to a platform specific API in your shared code. If you do, every project but the build output for the platform that implements the API will fail. So to get around that you have to use preprocessor stuff to make the code not appear at all in some outputs. Depending on what you are doing you may need to ignore entire classes EG

```c#
#if WINDOWS
public class MyWindowsSpecificClass : MyPlatformInterface {
....
}
#endif
```

```c#
MyPlatformInterface? val = null;
#if WINDOWS
val = new MyWindowsSpecifcClass
#endif

if (val != null) {
val.DoThings()
}
```

On non Windows outputs, the `MyWindowsSpecificClass` wont even exist. This however can cause other problems, as you also wont get any feedback in your IDE if one of the targets does not build properly until you explicitly switch to looking at it or build it. Because as far as your IDE is concerned, the code does not exist. You will also need to do stuff like that everytime you have a platform specific bit of functionality that uses a TFM specific API. As such, it can quickly grow into a very large, hard to parse, unmaintainable mess. My advice is not to go this way and instead use DI/runtime checks.

However if you use dependency injection you just have to declare an interface in some shared non platform specific project, and implement the interface in the platform specific project. Then inject all the platform specific services at runtime. Which is significantly simpler to manage and debug. There is an example here for implementing DI in a single project:

https://docs.avaloniaui.net/docs/guides/implementation-guides/how-to-implement-dependency-injection

I don't think there is an official guide for multi project DI, although you can use the one I wrote up here as a base (make sure you thread the entire thread, specifically the bit about not using the main method):

https://github.com/AvaloniaUI/Avalonia/discussions/13455#discussioncomment-7464329

**Reply by DmitryBorodiy** [reply]
I tried to do this, but for some reason it breaks Avalonia itself. After several tries I saw the error, anyway I'll try. Same approach on Uno Platform.

**Reply by DmitryBorodiy** [reply]
> > I just want to use Windows-specific APIs such as Share Dialog without losing cross-platform ability.
> 
> [FluentAvaloniaTheme](https://amwx.github.io/FluentAvaloniaDocs/pages/FATheme)

Yes, thank you! I know about that but this is only theme and no Windows specific API implementation.

**Reply by thevortexcloud** [reply]
What's the error?

**Reply by thevortexcloud** [reply]
You have a comma separated framework list. That is not valid. It should be something like:

```xml
<TargetFrameworks>net8.0;net8.0-windows</TargetFrameworks
```

**Comment by J0nathan550** [comment]
If you mean by WinUI3 UI, you have to use FluentAvaloniaTheme. It's available in Nuget Packages and has majority of all WinUI3 controls.

[FluentAvaloniaTheme](https://amwx.github.io/FluentAvaloniaDocs/pages/FATheme)

**Reply by DmitryBorodiy** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/57910623/8c12600c-0793-4896-81bc-7a5d1335d74a)

What's next? I just create new test project and change target frameworks. It's works with WPF but not works on Avalonia.

**Reply by thevortexcloud** [reply]
Avalonia does not support AnyCPU. You need to use x64/x86/what ever you are actually targeting.

**Reply by DmitryBorodiy** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/57910623/f0db912f-1c19-46bf-9f15-32d21105c03d)

**Reply by DmitryBorodiy** [reply]
After restarting the project everything worked

**Reply by DmitryBorodiy** [reply]
Some good news !

![image](https://github.com/AvaloniaUI/Avalonia/assets/57910623/2b0016fb-8ab6-4b0c-87fa-59ead29f4d19)

**Reply by jp2masa** [reply]
> Avalonia does not support AnyCPU. You need to use x64/x86/what ever you are actually targeting.

I don't think that's true. I read that `PlatformTarget` is practically deprecated and the target architecture is defined by RIDs now.

I can't find the reference right now, but [this SO answer](https://stackoverflow.com/a/68834371) points in the same direction.

**Reply by thevortexcloud** [reply]
I did try this the other day, but I was messing around with .NET Framework. So it probably does work different now. But it refused to work properly until I removed Any CPU. I do seem to vaguely recall also having issues with it on modern .NET too, but I have not tried it in a long time. It's mainly a problem due to the platform specific native libraries.

**Reply by jp2masa** [reply]
> I do seem to vaguely recall also having issues with it on modern .NET too, but I have not tried it in a long time. It's mainly a problem due to the platform specific native libraries.

Maybe if you have a .NET dependency that was not built for AnyCPU. But I'm pretty sure that loading native libraries is not limited by platform target on modern .NET, as I've loaded many different native libraries without ever using platform target. There may be some very specific libraries that work differently, though.

**Reply by maxkatz6** [reply]
Avalonia doesn't really care about PlatformTarget. But WinRT mappings might do.

**Reply by DmitryBorodiy** [reply]
https://github.com/microsoft/msix-packaging/issues/635

I have some problems with msix packaging and avalonia project which using target framework monikers.
